ffi/src/lib.rs

```diff
@@ -545,29 +1061,31 @@ pub unsafe extern "C" fn tx_change_list_entry_at(tx_report_list: *mut TxChangeLi
     Box::into_raw(report)
 }
 
+/// Returns the value at the provided `index` as a `Entid`.
+///
+/// # Panics
+///
+/// If there is no value present at the `index`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TransactionChange` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn changelist_entry_at(tx_report: *mut TransactionChange, index: c_int) -> Entid {
     let tx_report = &*tx_report;
     let index = index as usize;
     tx_report.changes[index].clone()
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn store_sync(store: *mut Store, user_uuid: *const c_char, server_uri: *const c_char) -> *mut ExternResult {
-    let store = &mut*store;
-    let user_uuid = c_char_to_string(user_uuid);
-    let server_uri = c_char_to_string(server_uri);
-    let res = store.sync(&server_uri, &user_uuid);
-    Box::into_raw(Box::new(res.into()))
-}
-
+/// destroy function for releasing the memory for `repr(C)` structs.
 #[no_mangle]
 pub unsafe extern "C" fn destroy(obj: *mut c_void) {
     if !obj.is_null() {
         let _ = Box::from_raw(obj);
     }
 }
 
+/// Creates a function with a given `$name` that releases the memroy for a type `$t`.
```

Comment:
A trait and a standard implementation would be nicer, I think.  Since we own the trait, we can define for all of these types coming from other crates; and we can handle `Vec<T>` generically, so we don't require specialization...

Reply:
I think this may be possible. Then I can define the `destroy` function only and call obj.destroy(). The only question I have is how possible it is to deal with generics over FFI, but I will experiment.

Reply:
Right, it turns out that you cannot `#[no_mangle]` a generic function. In order to use a generic in the function definition `#[no_mangle]` would have to be removed, which would make the function name very difficult to determine in order to call externally. 

Which means, in order to use the trait I would have to create a destroy function for each type anyway so I can call the trait function on that type.

All the ways I have tried to get around this result in a level of confusion that defeats the point of doing this.

Reply:
> All the ways I have tried to get around this result in a level of confusion that defeats the point of doing this.

Thanks, that's super valuable!  I'm happy to take this bit of wisdom and roll with it.  Perhaps in the future we'll learn something better.  Maybe a marker trait (like `Sync` or `Sized`) might be possible at some point.